head/src/main.rs

```diff
@@ -0,0 +1,138 @@
+}
+
+impl Flags {
+    /// Parse arguments into a Flags enum
+    ///
+    /// This will exit the program early on invalid args
+    fn parse(matches: &ArgMatches) -> Self {
+        match value_t!(matches, "bytes", usize) {
+            Ok(bytes_count) => Flags::BytesCount(bytes_count),
+            Err(e) => match e.kind {
+                ErrorKind::ValueValidation => e.exit(),
+                _ => match value_t!(matches, "lines", usize) {
+                    Ok(l) => Flags::LinesCount(l),
+                    Err(e) => match e.kind {
+                        ErrorKind::ValueValidation => e.exit(),
+                        _ => Flags::LinesCount(DEFAULT_LINES_COUNT),
+                    },
+                },
+            },
+        }
+    }
+}
+
+/// Input is either a list of one or more files, or STDIN
+enum Input {
+    Files(Vec<String>),
+    Stdin,
+}
+
+impl Input {
+    /// Parse arguments into an Input enum.
+    fn parse(matches: &ArgMatches) -> Self {
+        if let Some(files) = matches.values_of("FILE") {
+            Self::Files(files.map(|f| f.into()).collect())
+        } else {
+            Self::Stdin
+        }
+    }
+}
+
+/// Return the head of our input, truncated at a number of lines or bytes
+fn head(flags: Flags, input: Input) -> Result<(), io::Error> {
+    match input {
+        Input::Files(files) => {
+            let files_count = files.len();
+            for (i, file) in files.iter().enumerate() {
+                if files_count > 1 {
+                    if i != 0 {
+                        println!();
+                    }
+                    println!("==> {} <==", file);
+                }
+                let f = File::open(file)?;
+                let reader = BufReader::new(f);
+                read_stream(&flags, reader, &mut io::stdout())?;
+            }
+        }
+        Input::Stdin => {
+            let stdin = io::stdin();
+            let reader = BufReader::new(stdin.lock());
+            read_stream(&flags, reader, &mut io::stdout())?;
+        }
+    }
+    Ok(())
+}
+
+/// Read from a stream, truncated at a number of lines or bytes and write back to a stream
+fn read_stream<R: Read, W: Write>(
+    flags: &Flags, mut reader: BufReader<R>, writer: &mut W,
+) -> Result<(), io::Error> {
+    match flags {
+        Flags::LinesCount(lines_count) => {
+            for _ in 0..*lines_count {
+                let mut buffer = Vec::new();
+                let bytes_read = reader.read_until(NEW_LINE, &mut buffer)?;
+                if bytes_read == 0 {
+                    break;
+                }
+                writer.write_all(&buffer)?;
+            }
+        }
+        Flags::BytesCount(bytes_count) => {
+            let mut buffer = Vec::new();
+            reader.take(*bytes_count as u64).read_to_end(&mut buffer)?;
+            writer.write_all(&buffer)?;
+        }
+    }
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_lines() {
```

Comment:
The tests are ok, I would only make a more descriptive name for the test function.
It's always a good to have concise names with enough meaning on normal code, but on tests, is always a good idea to have very descriptive names, even when they get big. Like when running test it will only show the test function name and say it was ok or not, if the name is concise then maybe the meaning of the test may get ambiguous.

I would call 'read_stream_flags_linecount' and 'read_stream_flags_bytecount' or something similar